pkg/algorithm/cstorpoolselect/v1alpha1/select.go

```diff
@@ -103,6 +115,101 @@ type policy interface {
 	filter(*csp.CSPList) (*csp.CSPList, error)
 }
 
+// scheduleWithOverProvisioningAwareness is a pool
+// selection implementation.
+type scheduleWithOverProvisioningAwareness struct {
+	// overProvisioning field if true means over-provisioning is enabled or vice-versa.
+	overProvisioning bool
+	// spcName is the name of the SPC to which the over-provisioning policy will be
+	// applied and volume will be created from CSPs of this SPC.
+	spcName string
+	// openebsNamespace is the namespace where OpenEBS is installed.
+	openebsNamespace string
+	// totalCapacity is the capacity of the incoming volume.
+	totalCapacity resource.Quantity
+	// err constains a list of error in if any while building this current structure.
+	err []error
+}
+
+// priority returns the priority of the
+// policy implementation
+func (p scheduleWithOverProvisioningAwareness) priority() priority {
+	return highPriority
+}
+
+// name returns the name of the policy
+// implementation
+func (p scheduleWithOverProvisioningAwareness) name() policyName {
+	return overProvisioningPolicy
+}
+
+// filter selects the pools available on the host
+// for which the policy has been applied
+func (p scheduleWithOverProvisioningAwareness) filter(pools *csp.CSPList) (*csp.CSPList, error) {
+	if len(p.err) > 0 {
+		return nil, errors.Errorf("failed to fetch overprovisioning details:%v", p.err)
+	}
+
+	if p.overProvisioning {
+		klog.Infof("Overprovisioning restriction policy not added as overprovisioning is enabled on spc %s", p.spcName)
+		return pools, nil
+	}
+
+	filteredPools := &csp.CSPList{Items: []*csp.CSP{}}
+	for _, pool := range pools.Items {
+		volCap, err := p.consumedCapacity(pool.Object)
+		if err != nil {
+			klog.Errorf("failed to get capacity consumed by existing volumes on pool %s:{%s} ", pool.Object.UID, err.Error())
+			continue
+		}
+		if pool.HasSpace(p.totalCapacity, volCap) {
+			filteredPools.Items = append(filteredPools.Items, pool)
+		} else {
+			klog.V(2).Infof("Can't select CSP with UID %q: Required space not available: Policy %s", pool.Object.UID, overProvisioningPolicy)
+		}
+	}
+	return filteredPools, nil
+}
+
+// getAllVolumeCapacity returns the sum of total capacities of all the volumes
+// present of the given CSP.
+func (p *scheduleWithOverProvisioningAwareness) consumedCapacity(csp *apis.CStorPool) (resource.Quantity, error) {
+	var totalcapcity resource.Quantity
+	cstorVolumeMap := make(map[string]bool)
+	label := string(apis.CStorPoolKey) + "=" + csp.Name
+	cstorVolumeReplicaObjList, err := cstorvolumereplica.NewKubeclient().WithNamespace(p.openebsNamespace).List(metav1.ListOptions{LabelSelector: label})
```

Comment:
does CVRs always stay in openebsNamespace?

Reply:
`openebsNamespace` is the namespace where openebs is installed. And CVRs always stay in a namespace where openebs is installed.